### PR TITLE
new mapper api

### LINKED ISSUE #15: suggested urlMapper api
``` js
urlMapper.parse('/bar?bar=baz#hash', '/:foo'); // => { params: { foo: 'bar' }, query: { bar: 'baz' }, hash: 'hash' }
urlMapper.stringify('/:foo', { params: { foo: 'bar' }, query: { bar: 'baz' }, hash: 'hash' }); // => '/bar?bar=baz#hash'
urlMapper.map('/bar?bar=baz#hash', routes, function(match, values){
  match({ route: values })
  return 'some return value';
}); // => 'some return value'
urlMapper.map('/bar?bar=baz#hash', routes); // => { match, values }

```

I have plans to implement this api for next version of lib. Any suggestions?

### PR DESCRIPTION
**Pull request by Guria**
resolves #15 
@christianalfoni @maxfrigge ready for review. will update README and merge after review

### CODE REVIEW

**Comment by Guria** [comment]
@christianalfoni @maxfrigge updated README. Please review docs and spelling within it.

**Comment by maxfrigge** [comment]
Ahhm.. what would be the workflow for this? I thought I would fork the branch (which apparently is not possible) and then create a pull request :-/

**Comment by Guria** [comment]
Switch to changes tab and add inline comments with plus button

File: README.md [hunk 1]
```diff
@@ -1,28 +1,204 @@
...
+
+## Instalation
+
+`npm install url-mapper --save`
+
+## Usage
+
+### Overview
+
+The main purpose of `url-mapper` is match given `URL` to one of the `routes`.
```
Comment (maxfrigge) [inline README.md]: The main purpose of `url-mapper` is to match a given `URL` to one of the `routes`.

File: README.md [hunk 2]
```diff
@@ -1,28 +1,204 @@
...
+It will return matched route (key and associated value) and parsed parameters.
```
Comment (maxfrigge) [inline README.md]: It will return the matched route (key and associated value) and parsed parameters.

File: README.md [hunk 3]
```diff
@@ -1,28 +1,204 @@
...
+You can associate with route anything you want: function, React component or just plain object.
```
Comment (maxfrigge) [inline README.md]: Not quite sure on this one.. but I think it should be the other way around.

You can associate anything you want with route:

File: README.md [hunk 4]
```diff
@@ -1,28 +1,204 @@
...
+
+`url-mapper` is helpful when creating router packages for frameworks or can be used as router itself.
+It allows you to outsource work with url (mapping, parsing, stringifying) and concentrate on wiring up things related to your favorite framework.
```
Comment (maxfrigge) [inline README.md]: It allows you to outsource working with a url ..

File: README.md [hunk 5]
```diff
@@ -1,28 +1,204 @@
...
+  ReactDOM.render(
+    <CoreApp>
+      <Component {...props} />
+    </CoreApp>
+  );
+}
 ```
-**url-mapper** passes an object representing the parsed URL to your route callbacks. Given the URL `http://www.bigapp.com/foo/123?bar=baz` and a matching route `/foo/:id`, your callbacks will receive:
+
+See [`cerebral-router`](https://github.com/christianalfoni/cerebral-router) as example of building own router solution on top of `url-mapper`.
```
Comment (maxfrigge) [inline README.md]: .. as an example of building your own router solution on top of ..

File: README.md [hunk 6]
```diff
@@ -1,28 +1,204 @@
...
+
+## API
+
+### Main module
+
+At top level `url-mapper` module exports factory which returns default implementation of `URL` <==> `route(params)` converter.
```
Comment (maxfrigge) [inline README.md]: At the top level the `url-mapper` module exports a factory which returns the default implementation of a `URL` <==> `route(params)` converter.

File: README.md [hunk 7]
```diff
@@ -1,28 +1,204 @@
...
+  </tr>
+</table>
+
+##### Returns
+
+`Object` - Object with `parse`, `stringify` and `map` methods.
+
+Returned methods deals with Express-style route definitions and cleaned routable part of url (without origin, base path, leading hash symbol).
+
+Params defined in route is mapped to the same named properties in `values` Object with help of `path-to-regexp` module.
```
Comment (maxfrigge) [inline README.md]: Params defined in route are mapped to the same named properties in the `values` Object with help of `path-to-regexp` module.

File: README.md [hunk 8]
```diff
@@ -1,28 +1,204 @@
...
+It safe to pass Numbers and Booleans as well as Strings as path parameteres.
```
Comment (maxfrigge) [inline README.md]: It is safe ... parameters.

File: README.md [hunk 9]
```diff
@@ -1,28 +1,204 @@
...
+Original type would be preserved while parsing back stringified one.
```
Comment (maxfrigge) [inline README.md]: The original type would ...

File: README.md [hunk 10]
```diff
@@ -1,28 +1,204 @@
...
+
+By default, query part is ignored.
```
Comment (maxfrigge) [inline README.md]: .. the query part is ignored

File: README.md [hunk 11]
```diff
@@ -1,28 +1,204 @@
...
+Query part params is mapped to the same named properties in `values` Object if `{ query: true }` option was passed to factory.
```
Comment (maxfrigge) [inline README.md]: Query part params are mapped ...

File: README.md [hunk 12]
```diff
@@ -1,28 +1,204 @@
...
+Conversion for query part is made with help of `URLON` module. Therefore, it can accept any JSON serializable value.
```
Comment (maxfrigge) [inline README.md]: Conversion of the query part ..

File: README.md [hunk 13]
```diff
@@ -1,28 +1,204 @@
...
+------|----------|--------
+route | `String` | Express style route definition
+url   | `String` | Routable part of url
+
+##### Returns
+
+`Object` - values parsed from `url` with given `route`.
+
+Path parsed using `path-to-regexp` module, tweaked to support `Boolean` and `Number`.
+Query part parsed with `URLON` module if it enabled in factory.
```
Comment (maxfrigge) [inline README.md]: ... if `{ query: true }` option was passed to factory.

File: README.md [hunk 14]
```diff
@@ -1,28 +1,204 @@
...
+##### Usage
+
+`mapper.stringify(route, values)`;
+
+##### Arguments
+
+Param  | Type     | Details
+-------|----------|--------
+route  | `String` | Express style route definition
+values | `Object` | Object used to populate parameters to url
```
Comment (maxfrigge) [inline README.md]: Object used to populate parameters in route definition

File: README.md [hunk 15]
```diff
@@ -1,28 +1,204 @@
...
+
+##### Returns
+
+`String` - values stringified to `url` with given `route`.
+
+Properties defined at route is stringified to path part using `path-to-regexp` module, tweaked to support `Boolean` and `Number`.
```
Comment (maxfrigge) [inline README.md]: Properties defined in route are stringified to path part using..

File: README.md [hunk 16]
```diff
@@ -1,28 +1,204 @@
...
+Properties not defined at route is stringified to query part using `URLON` module if it enabled in factory.
```
Comment (maxfrigge) [inline README.md]: Properties not defined in route are stringified .. ... if { query: true } option was passed to factory.

File: README.md [hunk 17]
```diff
@@ -1,28 +1,204 @@
...
+url    | `String` | Routable part of url
+routes | `Object` | Routes to map url with
+
+##### Returns
+
+`Object` - Object representing matched route with properties:
+
+Property  | Type     | Details
+----------|----------|--------
+route     | `String` | Matched `route` defined as key at `routes`
```
Comment (maxfrigge) [inline README.md]: ... defined as key in `routes`

File: README.md [hunk 18]
```diff
@@ -1,28 +1,204 @@
...
+match     | `Any`    | Value from `routes` associated with matched `route`
+values    | `Object` | Values parsed from given `url` with matched `route`
+
+### Matcher
+
+Custom converting algoritms could be implemented by providing own compile function.
```
Comment (maxfrigge) [inline README.md]: .. by providing a custom compile function.

File: README.md [hunk 19]
```diff
@@ -1,28 +1,204 @@
...
+If you don't like default route definition format or converting algoritms, feel free to make your own.
```
Comment (maxfrigge) [inline README.md]: algorithm

File: README.md [hunk 20]
```diff
@@ -1,28 +1,204 @@
...
+
+##### Arguments
+
+Param     | Type       | Details
+----------|------------|--------
+compileFn | `Function` | Function used by mapper to "compile" a route.
+options   | `Any`      | `Optional`. Passed to `compileFn` as second argument.
+
+For each route mapper would call `compileFn(route, options)` and cache result internally.
+`compileFn` should return `parse(url)` and `stringify(values)` methods for given route.
```
Comment (maxfrigge) [inline README.md]: for any given route.

File: README.md [hunk 21]
```diff
@@ -1,28 +1,204 @@
...
+See [default implementation](/index.js#L6) for reference.
+
+##### Returns
+
+`Object` - Object with `parse(route, url)`, `stringify(route, values)` and `map(url, routes)` methods.
+
+This methods will use cached methods returned by `compileFn` for given routes.
```
Comment (maxfrigge) [inline README.md]: These methods ...

**Comment by maxfrigge** [comment]
I have kept it rather short and just put in what I would have written, which must not always be correct :)

**Comment by Guria** [comment]
@maxfrigge resolved your comments.
What do you say about overall doc? Does it clary enough?

File: README.md [hunk 22]
```diff
@@ -1,28 +1,204 @@
...
+
+##### Returns
+
+`Object` - Object with `parse`, `stringify` and `map` methods.
+
+Returned methods deals with Express-style route definitions and cleaned routable part of url (without origin, base path, leading hash symbol).
+
+Params defined in route are mapped to the same named properties in the `values` Object with help of `path-to-regexp` module.
+It is safe to pass Numbers and Booleans as well as Strings as path parameteres.
+The original type type would be preserved while parsing back stringified one.
```
Comment (maxfrigge) [inline README.md]: The original type ~~type~~ would

**Comment by maxfrigge** [comment]
I think it holds all the information required and should be ok as a starting point. For me personally it's too much text - I prefer shorter text with more code samples when reading docs, but that's a personal preference. 

I would say that as long as all the information required is present it should be good to go, there will always be time for another iteration, after some more feedback from other developers.

**Pull request merged**